Release 0 10 0 (#340)
kestra-io/kestra.io: content/docs/08.architecture.md

### Worker

The **Worker** is where all tasks and polling triggers are run. A Worker will receive a task from the Executor or a polling trigger from the Scheduler and will process it. As tasks and polling triggers can be nearly anything (heavy computations, simple API calls, etc.), the Worker is essentially a Thread Pool where you can configure the number of threads you need.

You can scale Workers as necessary and have many instances on multiple servers, each with its own Thread Pool.

As the worker executes the tasks and the polling triggers, it needs access to any external services they use (database, REST services, message broker, etc.). Workers are the only server components that need access to external services.

### Scheduler

> The **Scheduler** will handle most of the [Triggers](./05.developer-guide/08.triggers/index.md) (except the [Flow Triggers handled by the Executor](./05.developer-guide/08.triggers/02.flow.md)). It will continuously watch all triggers and, if all the conditions are met, will trigger an execution (submit it to the Executor).

Reviewer: ```suggestion
The **Scheduler** will handle most of the [Triggers](./05.developer-guide/08.triggers/index.md) except the [Flow Triggers handled by the Executor](./05.developer-guide/08.triggers/02.flow.md). It will continuously watch all triggers and, if all conditions are met, will execute the flow (submit the flow to the Executor).
```

triggers vs. "to trigger" as synonym to "run/execute" can be confusing

Author: The Scheduler will not execute the flow but trigger an execution.
The Flow is executed by the executor and the worker.

Maybe we can replace by '..., will start an execution (submit ...)

> In the case of polling triggers, the scheduler will decide if it must be evaluated (based on its evaluation interval), and if it must be, will send the polling trigger to the worker for evaluation. Polling triggers are non-reentrant: they cannot be evaluated concurrently, and they cannot be evaluated if an execution of the flow triggered by this trigger is already running.

Reviewer: ```suggestion
In the case of polling triggers, the scheduler will decide (based on the configured evaluation interval) whether to execute the flow. If the polling trigger conditions are met, the scheduler will send the execution, along with the trigger metadata, to the worker for execution. 

Note that polling triggers cannot be evaluated concurrently. They also can't be reevaluated i.e. if the execution of a flow, which started as a result of the same trigger, is already in a Running state, the next execution will not be started until the previous one finishes.
```
Reentrant is a bit of a confusing/not common word, seems easier to understand without it, up to you

Author: Yes, you basically explain what reentrancy is about so OK for me, better to be explicit.

> Internally, a Scheduler will tick every second and decide which trigger must evaluated.

Reviewer: ```suggestion
Internally, the Scheduler will keep checking every second whether some trigger must be evaluated.
```

As the worker executes the tasks and the polling triggers, it needs access to any external services they use (database, REST services, message broker, etc.). Workers are the only server components that need access to external services.

#### Worker Group (EE)

::alert{type="warning"}
This is an [Enterprise Edition](https://kestra.io/enterprise) feature.
::

A Worker Group is a set of workers that can be targeted specifically for a task execution or a polling trigger evaluation. For this, the task or the polling trigger must define the `workerGroup.key` property with the key of the worker group to target. A default worker group can also be configured at the namespace level.

> All tasks and polling triggers are executed on the un-named default worker group unless the `workerGroup.key` property is defined.

Reviewer: ```suggestion
if the `workerGroup.key` property is not provided, all tasks and polling triggers are executed on the default worker group. That default worker group doesn't have a dedicated key.
```

> Worker Groups can be used, for example, for the following needs:

Reviewer: ```suggestion
Here are common use cases in which Worker Groups can be beneficial:
```

- Execute tasks and polling triggers on specific computation capabilities (like GPU).

> - Execute tasks and polling triggers on specific OS.

Reviewer: ```suggestion
- Execute tasks and polling triggers on a worker with a specific Operating System (e.g., a Windows server).
```

- Restrict backend access to a set of workers (firewall rules, private networks, ...).
- Execute tasks and polling triggers close to a remote backend (region selection).

::alert{type="warning"}
Even if using worker groups, we strongly advise always having at least one worker in the un-named default worker group, as there is no way to enforce that all tasks and polling triggers have a `workerGroup.key` property set.
::

> There is currently no validation that a worker group exists, so a task or a polling trigger that defines a worker group that didn't exist will wait forever pending execution.

Reviewer: ```suggestion
There is currently no way to validate whether a worker group exists before using it in a task or trigger. If a task or a polling trigger defines a worker group that doesn't exist, it will wait forever, leaving the flow's Execution stuck in a pending state.
```
